File: Daily/2024-01-22/가장 많이 받은 선물.js
function solution(friends, gifts) {
  let dic = {};
  let l = friends.length;
  let arr = new Array(l).fill(0);
  friends.forEach(fr => {
    dic[fr] = [0, 0, {}];
    friends.forEach(f => {
      dic[fr][2][f] = 0;
    });
  });
  gifts.forEach(g => {
    let [a, b] = g.split(" ");
    dic[a][0]++;
    dic[b][1]++;
    dic[a][2][b]++;
  });

  for (let i = 0; i < l - 1; i++) {
    for (let j = i + 1; j < l; j++) {
      if (dic[friends[i]][2][friends[j]] || dic[friends[j]][2][friends[i]]) {
        let t = dic[friends[i]][2][friends[j]] - dic[friends[j]][2][friends[i]];
        if (t > 0) {
          arr[i]++;
        } else if (t < 0) {
          arr[j]++;
        } else {
          t = dic[friends[i]][0] - dic[friends[i]][1] - dic[friends[j]][0] + dic[friends[j]][1];
          if (t > 0) {
            arr[i]++;
          } else if (t < 0) {
            arr[j]++;
          }
        }
      } else {
        t = dic[friends[i]][0] - dic[friends[i]][1] - dic[friends[j]][0] + dic[friends[j]][1];
        if (t > 0) {
          arr[i]++;
        } else if (t < 0) {
          arr[j]++;
        }
      }
    }
  }
  return Math.max(...arr);
}

File: Daily/2024-01-22/가장 많이 받은 선물.py
def solution(friends, gifts):
    answer = 0
    dic = {}
    arr = [0] * len(friends)
    for i in friends:
        dic[i] = [0,0,{i:0 for i in friends}]
    for i in gifts:
        a,b = i.split()
        dic[a][0] += 1
        dic[b][1] += 1
        dic[a][2][b] += 1
    for i in range(len(friends)-1):
        for j in range(i+1,len(friends)):
            if dic[friends[i]][2][friends[j]] or dic[friends[j]][2][friends[i]]:
                t = dic[friends[i]][2][friends[j]] - dic[friends[j]][2][friends[i]]
                if t > 0:
                    arr[i] += 1
                elif t < 0:
                    arr[j] += 1
                else:
                    t = dic[friends[i]][0] - dic[friends[i]][1] - dic[friends[j]][0] + dic[friends[j]][1]
                    if t > 0:
                        arr[i] += 1
                    elif t < 0:
                        arr[j] += 1
            else:
                t = dic[friends[i]][0] - dic[friends[i]][1] - dic[friends[j]][0] + dic[friends[j]][1]
                if t > 0:
                    arr[i] += 1
                elif t < 0:
                    arr[j] += 1
                    
        
    return max(arr)
